app/lib/shared/cached_value.dart

```diff
@@ -0,0 +1,75 @@
+// Copyright (c) 2021, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:logging/logging.dart';
+import 'package:meta/meta.dart';
+import 'package:retry/retry.dart';
+
+final _logger = Logger('cached_value');
+
+typedef UpdateFn<T> = Future<T> Function();
+
+/// A locally cached and periodically updated value that can be accessed
+/// synchronously.
+class CachedValue<T> {
+  final UpdateFn<T> _updateFn;
+  final RetryOptions _retryOptions;
+  Timer _timer;
+  T _value;
+  Completer _ongoingCompleter;
+  bool _initialized = false;
+
+  CachedValue({
+    @required UpdateFn<T> fn,
+    RetryOptions retryOptions,
+  })  : _updateFn = fn,
+        _retryOptions = retryOptions ??
+            RetryOptions(
+              delayFactor: Duration(seconds: 3),
+              maxAttempts: 3,
+            );
+
+  bool get isInitialized => _initialized;
+
+  T get value => _value;
+  set value(T v) {
+    _value = v;
+    _initialized = true;
+  }
+
+  /// Updates the cached value.
+  Future<void> update() async {
+    if (_ongoingCompleter != null) {
+      await _ongoingCompleter.future;
+      return;
+    }
+    _ongoingCompleter = Completer();
+    try {
+      await _retryOptions.retry(() async {
```

Comment:
Should we not avoid retrying here... and instead document that `updateFn` should retry as needed.